flagd-proxy/tests/loadtest/pkg/handler/handler.go

```diff
@@ -0,0 +1,179 @@
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	"buf.build/gen/go/open-feature/flagd/grpc/go/sync/v1/syncv1grpc"
+	"github.com/open-feature/flagd/flagd-proxy/tests/loadtest/pkg/client"
+	trigger "github.com/open-feature/flagd/flagd-proxy/tests/loadtest/pkg/trigger"
+	"github.com/open-feature/flagd/flagd-proxy/tests/loadtest/pkg/watcher"
+)
+
+type Handler struct {
+	config  Config
+	trigger trigger.Trigger
+}
+
+type Config struct {
+	FilePath string `json:"filePath"`
+	Host     string `json:"host"`
+	Port     uint16 `json:"port"`
+	OutFile  string `json:"outFile"`
+}
+
+type TestConfig struct {
+	Watchers int           `json:"watchers"`
+	Repeats  int           `json:"repeats"`
+	Delay    time.Duration `json:"delay"`
+}
+
+type TestResult struct {
+	TotalTime      time.Duration `json:"totalTime"`
+	TimePerWatcher time.Duration `json:"timePerWatcher"`
+}
+
+type ProfilingResults struct {
+	Tests                 []TestResult  `json:"tests"`
+	AverageTotalDuration  time.Duration `json:"averageTotalDuration"`
+	AverageTimePerWatcher time.Duration `json:"averageTimePerWatcher"`
+	Watchers              int           `json:"watchers"`
+	Repeats               int           `json:"repeats"`
+}
+
+func NewHandler(config Config, trigger trigger.Trigger) *Handler {
+	return &Handler{
+		config:  config,
+		trigger: trigger,
+	}
+}
+
+func (h *Handler) Profile(ctx context.Context, configs []TestConfig) ([]ProfilingResults, error) {
+	out := []ProfilingResults{}
+	for _, config := range configs {
+		if err := h.trigger.Setup(); err != nil {
+			return []ProfilingResults{}, err
+		}
+		results := []TestResult{}
+		for i := 1; i <= config.Repeats; i++ {
+			fmt.Printf("starting profile %d\n", i)
+			res := h.runTest(ctx, config.Watchers)
+			results = append(results, res)
+			fmt.Println("-----------------------")
+			time.Sleep(config.Delay)
+		}
+		timePer := time.Duration(0)
+		totalTime := time.Duration(0)
+		for _, res := range results {
+			timePer += res.TimePerWatcher
+			totalTime += res.TotalTime
+		}
+		out = append(out, ProfilingResults{
+			Watchers:              config.Watchers,
+			Repeats:               config.Repeats,
+			Tests:                 results,
+			AverageTotalDuration:  totalTime / time.Duration(config.Repeats),
+			AverageTimePerWatcher: timePer / time.Duration(config.Repeats),
+		})
+	}
+
+	return out, h.writeFile(out)
+}
+
+//nolint:funlen
+func (h *Handler) runTest(ctx context.Context, watchers int) TestResult {
+	readyWg := sync.WaitGroup{}
+	readyWg.Add(watchers)
+	readySuccess := make(chan bool, 1)
+
+	finishedWg := sync.WaitGroup{}
+	finishedWg.Add(watchers)
+	finishedSuccess := make(chan bool, 1)
+
+	errChan := make(chan error, 1)
+
+	fmt.Printf("starting %d watchers...\n", watchers)
+	var c syncv1grpc.FlagSyncServiceClient
+	var err error
+
+	for i := 0; i < watchers; i++ {
+		if i%250 == 0 {
```

Comment:
I assume we are having connection issues when reaching ~250 connections ?

Reply:
yes this is the default maximum number of streams per client, this can also be enforced on the client side, but I felt it safer to stay within the default bounds for this tool 
https://github.com/golang/net/blob/9001ca7de9d738aed247bf005debccb5b186a965/http2/server.go#L58